modules/nuxeo-apidoc-core/src/main/java/org/nuxeo/apidoc/export/graphs/introspection/AbstractGraphExporter.java

```diff
@@ -0,0 +1,391 @@
+                            NODE_TYPE.COMPONENT.name());
+                    graph.addEdge(createEdge(compNode, refNode, EDGE_TYPE.SOFT_REQUIRES.name()));
+                }
+
+            }
+        }
+
+        // Adds potentially missing bundle root
+        String rrId = getBundleRootId();
+        Node<?> rrNode = graph.getNode(rrId);
+        if (rrNode == null) {
+            rrNode = createBundleRoot();
+            graph.addNode(rrNode);
+        }
+        // Adds potentially missing link to bundle root for isolated (children nodes)
+        List<Node<?>> orphans = new ArrayList<>();
+        // handle missing references in all edges too
+        for (Edge edge : graph.getEdges()) {
+            Node<?> source = graph.getNode(edge.getSource());
+            if (source == null) {
+                source = addMissingNode(graph, edge.getSource());
+            }
+            Node<?> target = graph.getNode(edge.getTarget());
+            if (target == null) {
+                target = addMissingNode(graph, edge.getTarget());
+                if (NODE_TYPE.BUNDLE.name().equals(target.getType())
+                        && EDGE_TYPE.REQUIRES.name().equals(edge.getValue())) {
+                    orphans.add(target);
+                }
+            }
+        }
+        // handle isolated bundles
+        for (Node<?> orphan : orphans) {
+            requireBundleRootNode(graph, rrNode, orphan);
+        }
+        for (String child : isolatedBundles) {
+            requireBundleRootNode(graph, rrNode, graph.getNode(child));
+        }
+
+        // handle packages
+        for (PackageInfo pkg : distribution.getPackages()) {
+            if (filter != null && !filter.accept(pkg)) {
+                continue;
+            }
+            Node<?> pkgNode = createPackageNode(pkg, NODE_CATEGORY.PLATFORM);
+            graph.addNode(pkgNode);
+            for (String bundleId : pkg.getBundles()) {
+                Node<?> refNode = createReferenceNode(NODE_TYPE.BUNDLE.prefix(bundleId), NODE_TYPE.BUNDLE.name());
+                graph.addEdge(createEdge(pkgNode, refNode, EDGE_TYPE.CONTAINS.name()));
+                hit(nodeHits, pkgNode.getId());
+            }
+        }
+
+        refine(graph, nodeHits);
+
+        return graph;
+    }
+
+    /**
+     * Introspect the graph, centered on services.
+     * <p>
+     * select only services (resolving their dependencies based on their bundle's), linking them as parent of their
+     * extension points, and handling their contributions as a different relation.
+     * <p>
+     * When more than one service is declared within the same component, corresponding extension points and
+     * contributions will be "shared".
+     * <p>
+     * Other contributions, that cannot be linked to any service within the same components, are handled as mere
+     * "contributions" and are only linked to the extension point that they're contributing to.
+     */
+    // TODO: not used
```

Comment:
Forgotten TODO?

Reply:
actually part of the WIP..., will move all this to a later PR :+1: